### PR TITLE
[#3313] feat(common): refactor JsonUtils object mapper to client and service side

### PR DESCRIPTION
**Pull request by FANNG1**
### What changes were proposed in this pull request?
1. provide separate method to get object mapper in  client and service side
2. rename `JsonUtils.objectMapper` to `JsonUtils.objectMapperForTest`

### Why are the changes needed?

Fix: #3313 

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
existing tests

### CODE REVIEW

**Comment by FANNG1** [comment]
@yuqi1129 @diqiu50 , could you help to review this PR when you have time, thanks

File: common/src/main/java/com/datastrato/gravitino/json/JsonUtils.java [hunk 1]
```diff
@@ -236,9 +237,14 @@ private static class ObjectMapperHolder {
...
-   * @return The ObjectMapper instance.
+   * <p>This instance should only be used for testing. For production or other environments, obtain
+   * an {@link ObjectMapper} from the appropriate {@code ObjectMapperProvider} on the server or
+   * client side.
+   *
+   * @return the shared {@link ObjectMapper} instance for tests.
    */
-  public static ObjectMapper objectMapper() {
+  @VisibleForTesting
+  public static ObjectMapper objectMapperForTest() {
```
Comment (yuqi1129) [inline common/src/main/java/com/datastrato/gravitino/json/JsonUtils.java:247]: If it's only for testing,  `@VisibleForTesting` seems to be enough, the name `objectMapperForTest` is redundant.

Comment (FANNG1) [reply common/src/main/java/com/datastrato/gravitino/json/JsonUtils.java:247]: ok

Comment (FANNG1) [reply common/src/main/java/com/datastrato/gravitino/json/JsonUtils.java]: done

**Comment by FANNG1** [comment]
@yuqi1129 could you help to review again?

**Comment by FANNG1** [comment]
@yuqi1129 do you have any more comments?

**Comment by yuqi1129** [comment]
> @yuqi1129 do you have any more comments?

No, I'm okay with the current changes.

**Comment by FANNG1** [comment]
@jerryshao do you have time to review this?

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/ObjectMapperProvider.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.fasterxml.jackson.databind.cfg.EnumFeature;
+import com.fasterxml.jackson.databind.json.JsonMapper;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+
+/**
+ * Provides a singleton {@link ObjectMapper} configured for specific serialization and
+ * deserialization behaviors.
+ */
+public class ObjectMapperProvider {
```
Comment (jerryshao) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/ObjectMapperProvider.java:20]: Can you please add a private constructor?

File: server/src/main/java/com/datastrato/gravitino/server/web/ObjectMapperProvider.java [hunk 3]
```diff
@@ -4,20 +4,46 @@
...
+            .enable(MapperFeature.ACCEPT_CASE_INSENSITIVE_ENUMS)
+            .build()
+            .setSerializationInclusion(JsonInclude.Include.NON_NULL)
+            .registerModule(new JavaTimeModule());
+  }
+
+  /**
+   * Retrieves a globally shared {@link ObjectMapper} instance.
+   *
+   * <p>Note: This ObjectMapper is shared across all users. If you need to modify the default
```
Comment (jerryshao) [inline server/src/main/java/com/datastrato/gravitino/server/web/ObjectMapperProvider.java:34]: `shared across all users`, I think it is not accurate for server-side object mapper, you need to improve the comment here.

File: common/src/main/java/com/datastrato/gravitino/json/JsonUtils.java [hunk 4]
```diff
@@ -237,8 +238,13 @@ private static class ObjectMapperHolder {
   /**
    * Get the shared ObjectMapper instance for JSON serialization/deserialization.
    *
-   * @return The ObjectMapper instance.
+   * <p>This instance should only be used for testing. For production or other environments, obtain
+   * an {@link ObjectMapper} from the appropriate {@code ObjectMapperProvider} on the server or
+   * client side.
+   *
+   * @return the shared {@link ObjectMapper} instance for tests.
```
Comment (jerryshao) [inline common/src/main/java/com/datastrato/gravitino/json/JsonUtils.java:245]: I think your comments should be differentiated between server and client. Currently, it is the same.

**Comment by jerryshao** [comment]
We don't have to backport this to branch-0.5, seems like this is just a refractory. WDYT? @FANNG1

**Comment by yuqi1129** [comment]
> We don't have to backport this to branch-0.5, seems like this is just a refractory. WDYT? @FANNG1

As you mentioned, there is no need to cherry-pick such PRs as it's only a minor refactor or improvement, what concerns me is that it may cause potential code conflicts in the future.

File: common/src/main/java/com/datastrato/gravitino/json/JsonUtils.java (hunk 4 above)
Comment (FANNG1) [reply common/src/main/java/com/datastrato/gravitino/json/JsonUtils.java]: fixed

**Comment by jerryshao** [comment]
I'm more worried about the potential unknown issues introduced by the refactoring. As 0.5.1 is a stable release, maybe we should not include such refactoring PR avoid the issues.

File: server/src/main/java/com/datastrato/gravitino/server/web/ObjectMapperProvider.java (hunk 3 above)
Comment (FANNG1) [reply server/src/main/java/com/datastrato/gravitino/server/web/ObjectMapperProvider.java]: done

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/ObjectMapperProvider.java (hunk 2 above)
Comment (FANNG1) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/ObjectMapperProvider.java:20]: done

**Comment by FANNG1** [comment]
@jerryshao @yuqi1129 , this PR blocks running spark multi-version integrate test in embedded mode #3305 . so it needs to backport 0.5.1

**Review by jerryshao — APPROVED** [review]

**Pull request merged**